src/main/extractZipInplace.js

```diff
@@ -0,0 +1,55 @@
+import fs from 'fs';
+import path from 'path';
+
+import yauzl from 'yauzl';
+
+import { getLogger } from '../logger';
+
+const logger = getLogger(__filename.split('/').slice(-1)[0]);
+
+export default function extractZipInplace(zipFilePath) {
+  return new Promise((resolve) => {
+    const extractToDir = path.dirname(zipFilePath);
+    logger.info(`extracting ${zipFilePath}`);
+    const options = {
+      lazyEntries: true,
+      autoClose: true,
+    };
+    yauzl.open(zipFilePath, options, (error, zipfile) => {
+      if (error) throw error;
+      zipfile.on('entry', (entry) => {
+        const writePath = path.join(extractToDir, entry.fileName);
+        // if entry is a directory
+        if (/\/$/.test(entry.fileName)) {
+          fs.mkdir(writePath, (e) => {
+            if (e) {
+              if (e.code === 'EEXIST') { } else throw e;
+            }
+            zipfile.readEntry();
+          });
+        } else {
+          zipfile.openReadStream(entry, (err, readStream) => {
+            if (err) throw err;
+            readStream.on('end', () => {
+              zipfile.readEntry();
+            });
+            // Sometimes an entry will be in a dir, where the
+            // dir itself was *not* an entry, therefore we still need
+            // to create the dir here.
+            fs.mkdir(path.dirname(writePath), (e) => {
+              if (e) {
+                if (e.code === 'EEXIST') { } else throw e;
+              }
+              const writable = fs.createWriteStream(writePath);
+              readStream.pipe(writable);
+            });
+          });
+        }
+      });
+      zipfile.on('close', () => {
+        resolve(true);
+      });
+      zipfile.readEntry();
```

Comment:
I agree. I know such a thing is not in `yauzl`. It might exist elsewhere and I'm open to switching. Programming this taught me a bit about read/writeStreams and promises, so no regrets!

For our use case, we also want to make sure the dir structure of the output matches the zip archive, which may or may not be true for some higher-level library. Especially because I think sometimes the directories themselves are "entries" in the archive and sometimes they are not (even when the archived files have the dir in their path). So it took a bit of custom logic to make sure the output dir is always created.

Reply:
I read that thread you linked above and it's too bad there isn't one stable, generally-agreed-upon unzip npm module. Could you remind me what we're downloading? When would the output structure not match the archive? If you can accomplish this with `unzip` or another shell command, maybe calling that from `child_process` would be more straightfoward

Reply:
Yep, these are the invest sample data downloads. So they're folders with varying levels of subfolders. 

`unzip` would be nice, but I hesitate to depend on a program that we're not actually distributing with the app. Windows users won't have `unzip`, for example.

Reply:
@emlys to see an example archive that would not necessarily extract to the same output structure without all this logic, check out the test we have for this, where the directories themselves are _not_ "entries" in the archive. https://github.com/natcap/invest-workbench/pull/147/files#diff-84796f464c3e43389936aa12df7ce439dab4dc511f7bc43e7143cd03be25197aR107

Of course if we do find a nice higher-level utility for this, I would hope it handles this case by default. I assume `unzip` does.